### PR TITLE
Add RMSNorm module

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Similar to https://github.com/facebookresearch/multimodal/blob/dbeed9724bc9099be173c86871df7d41f3b7e58c/torchmultimodal/modules/layers/normalizations.py#L51

**The implementation here is not optimized and we welcome pull requests to improve this**

- Use `normalized_shape` instead of singular integer `dim` to be aligned with the `nn.LayerNorm` implementation
- Remove the [upcast to float and downcast 
](https://github.com/facebookresearch/multimodal/blob/dbeed9724bc9099be173c86871df7d41f3b7e58c/torchmultimodal/modules/layers/normalizations.py#L73)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121364

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D55485840](https://our.internmc.facebook.com/intern/diff/D55485840)

### CODE REVIEW

**Comment by wanchaol** [comment]
Yay love to see this!

**Comment by wanchaol** [comment]
cc @stas00 finally your ask is here!

**Comment by stas00** [comment]
thank you for the heads up, @wanchaol!

Do you know if it is on par with `apex`'s version?

File: torch/nn/functional.py [hunk 1]
```diff
@@ -2572,6 +2572,25 @@ def layer_norm(
...
+) -> Tensor:
+    r"""Apply Root Mean Square Layer Normalization.
+
+    See :class:`~torch.nn.RMSNorm` for details.
+    """
+    if has_torch_function_variadic(input, weight):
+        return handle_torch_function(
+            rms_norm, (input, weight), input, normalized_shape, weight=weight, eps=eps
+        )
+    if input.shape[-len(normalized_shape):] != normalized_shape:
```
Comment (albanD) [inline torch/nn/functional.py:2589]: If this error checking is here, does it mean that calling torch.rms_norm() without these will error out? Should we move these checks to c++ ?

File: torch/nn/modules/normalization.py [hunk 2]
```diff
@@ -292,6 +292,73 @@ def extra_repr(self) -> str:
...
+        normalized_shape (int or list or torch.Size): input shape from an expected input
+            of size
+
+            .. math::
+                [* \times \text{normalized\\_shape}[0] \times \text{normalized\\_shape}[1]
+                    \times \\ldots \times \text{normalized\\_shape}[-1]]
+
+            If a single integer is used, it is treated as a singleton list, and this module will
+            normalize over the last dimension which is expected to be of that specific size.
+        eps: a value added to the denominator for numerical stability. Default: 1e-6
```
Comment (albanD) [inline torch/nn/modules/normalization.py:319]: We should have eps be dtype-dependent. I think it's an oversight on our older modules where we assume they would always run in fp32.
I think we should start using `torch.finfo(x.dtype).eps` here WDYT?
cc @ngimel

**Review by albanD — COMMENTED** [review]
Sounds pretty good!
Feel free to keep autograd and more optimized kernels for the next PR if you want to keep this one simple.

**Comment by wanchaol** [comment]
> Do you know if it is on par with `apex`'s version?

@stas00 This seems to be the initial version adding the initial support of RMSNorm module that calls into basic aten op in C++, but looks like @mikaylagawarecki @albanD have plans to add more optimized kernels too!

File: docs/source/conf.py [hunk 3]
```diff
@@ -3023,6 +3023,7 @@
     "GroupNorm",
     "LayerNorm",
     "LocalResponseNorm",
+    "RMSNorm",
```
Comment (albanD) [inline docs/source/conf.py:3026]: Given that you updated the .rst files, I expect this is not needed?
Also adding new entry here is not allowed!

File: torch/nn/functional.py [hunk 4]
```diff
@@ -2572,6 +2572,24 @@ def layer_norm(
...
+    normalized_shape: List[int],
+    weight: Optional[Tensor] = None,
+    eps: Optional[float] = None,
+) -> Tensor:
+    r"""Apply Root Mean Square Layer Normalization.
+
+    See :class:`~torch.nn.RMSNorm` for details.
+    """
+    # isintance(input, Tensor) needed for classes that implement __torch_function__ but are not Tensors
+    if eps is None and isinstance(input, Tensor):
```
Comment (albanD) [inline torch/nn/functional.py:2586]: I'm curious why we resolve this here and not after torch_function handling or within the C++ code?

**Review by albanD — COMMENTED** [review]
Very cool!
Only two small things!

File: docs/source/conf.py (hunk 3 above)
Comment (mikaylagawarecki) [reply docs/source/conf.py:3026]: If I do not add this here it complains about "forward", "__repr__" and "reset_parameters" in `torch.nn.normalizations` not being documented, but none of the other nn modules document this

Comment (albanD) [reply docs/source/conf.py:3025]: `__repr__` sounds suprising?
reset_parameters should most likely be documented? forward would make sense as well.

In any case, we have a pretty strict rule that nothing should be added here I'm afraid.

Comment (mikaylagawarecki) [reply docs/source/conf.py:3025]: Sorry it was `extra_repr` not `__repr__`

This is the output of coverage.txt
```
torch.nn.modules.normalization
------------------------------
Classes:
 * RMSNorm -- missing methods:

   - extra_repr
   - forward
   - reset_parameters
   ```
   
   If I uncomment any nn.Module in docs/source/conf.py they will have similar errors.
   
   
   Does this mean you want us to document `torch.nn.modules.normalization`? Removing this line will make the error go away
   
https://github.com/pytorch/pytorch/blob/3db64c19554d644d09d2a16931bae7ebe96087fc/docs/source/nn.rst?plain=1#L556

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by stas00** [comment]
Thank you for this long awaited addition, @mikaylagawarecki 

Is this going to appear in 2.3.0?

**Comment by mikaylagawarecki** [comment]
@stas00 The 2.3 branch cut happened a couple weeks ago so this will not appear in 2.3 :/

**Comment by stas00** [comment]
Thank you, @mikaylagawarecki 

so 2.3.1 or 2.4?

Also is this an equivalent of `apex.normalization.FusedRMSNorm` performance wise? https://github.com/pytorch/pytorch/issues/72643

**Comment by mikaylagawarecki** [comment]
@stas00 This PR only provides scaffolding and an unoptimized implementation for the module, and will be available in 2.4

**Comment by jbschlosser** [comment]
With the scaffolding in place, we'd accept a PR backing it with a fused implementation for perf when applicable.

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55485840).

**Review by albanD — APPROVED** [review]
Thanks!
